torchtune/modules/transforms/tokenizers/_hf_tokenizer.py

```diff
@@ -146,3 +151,158 @@ def decode(self, token_ids: list[int]) -> str:
+    the specific ModelTokenizer and custom PromptTemplate.
+
+    Args:
+        tokenizer_json_path (str): Path to tokenizer.json file
+        tokenizer_config_json_path (Optional[str]): Path to tokenizer_config.json file. Default: None
+        generation_config_path (Optional[str]): Path to generation_config.json file.
+            Default: None
+        truncation_type (str): type of truncation to apply, either "left" or "right".
+            Default is "right".
+    """
+
+    def __init__(
+        self,
+        tokenizer_json_path: str,
+        *,
+        tokenizer_config_json_path: Optional[str] = None,
+        generation_config_path: Optional[str] = None,
+        truncation_type: str = "right",
+    ):
+        self.base_tokenizer = HuggingFaceBaseTokenizer(
+            tokenizer_json_path=tokenizer_json_path,
+            tokenizer_config_json_path=tokenizer_config_json_path,
+            generation_config_path=generation_config_path,
+        )
+
+        # Contents of the tokenizer_config.json
+        config = self.base_tokenizer.config
+
+        self.special_tokens = _infer_special_tokens_from_hf_config(config)
+        self.top_level_variables = self.extract_top_level_variables(config)
+
+        _env = jinja2.Environment(undefined=StrictUndefined)
+
+        # It is used sometimes in HF chat_templates
+        _env.globals["raise_exception"] = self._raise_helper
+
+        self.template = _env.from_string(config["chat_template"])
+        self.truncation_type = truncation_type
+
+        self.special_tokens_mapping = {}
+        for token in self.special_tokens:
+            self.special_tokens_mapping[token] = self.base_tokenizer.encode(token)
+
+    def _raise_helper(self, message: str):
+        raise jinja2.exceptions.TemplateError(message)
+
+    def extract_top_level_variables(self, config):
+        top_level = {}
+        for key, value in config.items():
+            if not isinstance(value, (dict, list)):
+                top_level[key] = value
+        return top_level
+
+    def tokenize_messages(
+        self,
+        messages: list[Message],
+        add_eos: bool = True,
+        max_seq_len: Optional[int] = None,
+    ) -> tuple[list[int], list[bool]]:
+        tokenized_messages = []
+        mask = []
+        previous_tokens = []
+
+        for i, message in enumerate(messages):
+            current_messages = [
+                {"role": m.role, "content": m.content[0]["content"]}
+                for m in messages[: i + 1]
+            ]
+
+            rendered = self.template.render(
+                messages=current_messages,
+                add_generation_prompt=add_eos if i == len(messages) - 1 else False,
+                **self.special_tokens_mapping,  # We assume that the naming is consistent
+                **self.top_level_variables,
+            )
+
+            current_tokens = self.base_tokenizer.encode(rendered, add_eos=False)
```

Comment:
Regarding the duplicate bos tokens, I suspect we just need to add an additional case [here](https://github.com/pytorch/torchtune/blob/20bdf1086e947fd010dd79faf1c78600062cb162/torchtune/modules/transforms/tokenizers/_hf_tokenizer.py#L132-L133) to deal with the fact that we are often passing a string like "<bos>..." to the base_tokenizer's encode method